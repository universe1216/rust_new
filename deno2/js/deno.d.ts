// Copyright 2018 Ryan Dahl <ry@tinyclouds.org>
// All rights reserved. MIT License.
type MessageCallback = (msg: ArrayBuffer) => void;
declare function denoSub(channel: string, cb: MessageCallback): void;
declare function denoPub(channel: string, msg: ArrayBuffer): null | ArrayBuffer;
declare function denoPrint(x: string): void;
