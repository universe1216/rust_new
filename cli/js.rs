// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.

use deno_core::Snapshot;

static CLI_SNAPSHOT: &[u8] =
  include_bytes!(concat!(env!("OUT_DIR"), "/CLI_SNAPSHOT.bin"));

pub fn deno_isolate_init() -> Snapshot {
  Snapshot::Static(CLI_SNAPSHOT)
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn runtime_snapshot() {
    let mut js_runtime = deno_core::JsRuntime::new(deno_core::RuntimeOptions {
      startup_snapshot: Some(deno_isolate_init()),
      ..Default::default()
    });
    js_runtime
      .execute_script(
        "<anon>",
        r#"
      if (!(bootstrap.mainRuntime && bootstrap.workerRuntime)) {
        throw Error("bad");
      }
      console.log("we have console.log!!!");
    "#,
      )
      .unwrap();
  }
}
