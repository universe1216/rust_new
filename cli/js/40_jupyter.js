// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.

const core = globalThis.Deno.core;
const internals = globalThis.__bootstrap.internals;

function enableJupyter() {
  const {
    op_jupyter_broadcast,
  } = core.ensureFastOps();

  globalThis.Deno.jupyter = {
    async broadcast(msgType, content, { metadata = {}, buffers = [] } = {}) {
      await op_jupyter_broadcast(msgType, content, metadata, buffers);
    },
  };
}

internals.enableJupyter = enableJupyter;
