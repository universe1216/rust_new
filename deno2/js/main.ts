/// <reference path="deno.d.ts" />
import { deno as pb } from "./msg.pb";
import * as ts from "typescript";

const globalEval = eval;
const window = globalEval("this");

window["denoMain"] = () => {
  denoPrint(`ts.version: ${ts.version}`);
  const res = denoPub("startDeno2", emptyArrayBuffer());
  //denoPrint(`after`);
  const resUi8 = new Uint8Array(res);
  denoPrint(`before`);
  const msg = pb.Msg.decode(resUi8);
  denoPrint(`after`);
  const {
    startCwd: cwd,
    startArgv: argv,
    startDebugFlag: debugFlag,
    startMainJs: mainJs,
    startMainMap: mainMap
  } = msg;

  denoPrint(`cwd: ${cwd}`);
  denoPrint(`debugFlag: ${debugFlag}`);

  for (let i = 0; i < argv.length; i++) {
    denoPrint(`argv[${i}] ${argv[i]}`);
  }
};

function typedArrayToArrayBuffer(ta: Uint8Array): ArrayBuffer {
  return ta.buffer.slice(
    ta.byteOffset,
    ta.byteOffset + ta.byteLength
  ) as ArrayBuffer;
}

function emptyArrayBuffer(): ArrayBuffer {
  return typedArrayToArrayBuffer(new Uint8Array([]));
}
