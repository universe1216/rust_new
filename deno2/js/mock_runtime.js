// A simple runtime that doesn't involve typescript or protobufs to test
// libdeno. Invoked by mock_runtime_test.cc

function assert(cond) {
  if (!cond) throw Error("mock_runtime.js assert failed");
}

function typedArrayToArrayBuffer(ta) {
  return ta.buffer.slice(ta.byteOffset, ta.byteOffset + ta.byteLength);
}

function CanCallFunction() {
  deno.print("Hello world from foo");
  return "foo";
}

// This object is created to test snapshotting.
// See DeserializeInternalFieldsCallback and SerializeInternalFieldsCallback.
const snapshotted = new Uint8Array([1, 3, 3, 7]);

function TypedArraySnapshots() {
  assert(snapshotted[0] === 1);
  assert(snapshotted[1] === 3);
  assert(snapshotted[2] === 3);
  assert(snapshotted[3] === 7);
}

function PubSuccess() {
  deno.sub((channel, msg) => {
    assert(channel === "PubSuccess");
    deno.print("PubSuccess: ok");
  });
}

function PubByteLength() {
  deno.sub((channel, msg) => {
    assert(channel === "PubByteLength");
    assert(msg instanceof ArrayBuffer);
    assert(msg.byteLength === 3);
  });
}

function SubReturnEmpty() {
  const ui8 = new Uint8Array("abc".split("").map(c => c.charCodeAt(0)));
  const ab = typedArrayToArrayBuffer(ui8);
  let r = deno.pub("SubReturnEmpty", ab);
  assert(r == null);
  r = deno.pub("SubReturnEmpty", ab);
  assert(r == null);
}

function SubReturnBar() {
  const ui8 = new Uint8Array("abc".split("").map(c => c.charCodeAt(0)));
  const ab = typedArrayToArrayBuffer(ui8);
  const r = deno.pub("SubReturnBar", ab);
  assert(r instanceof ArrayBuffer);
  assert(r.byteLength === 3);
  const rui8 = new Uint8Array(r);
  const rstr = String.fromCharCode(...rui8);
  assert(rstr === "bar");
}

function DoubleSubFails() {
  // deno.sub is an internal function and should only be called once from the
  // runtime.
  deno.sub((channel, msg) => assert(false));
  deno.sub((channel, msg) => assert(false));
}

// The following join has caused SnapshotBug to segfault when using kKeep.
[].join("");

function SnapshotBug() {
  assert("1,2,3" === String([1, 2, 3]));
}
