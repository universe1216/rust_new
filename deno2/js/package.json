{
  "devDependencies": {
    "parcel-bundler": "^1.8.1",
    "protobufjs": "^6.8.6",
    "typescript": "^2.9.1"
  }
}
