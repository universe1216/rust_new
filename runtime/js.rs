// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
#[cfg(not(feature = "dont_create_runtime_snapshot"))]
use deno_core::Snapshot;
#[cfg(not(feature = "dont_create_runtime_snapshot"))]
use log::debug;
#[cfg(not(feature = "dont_create_runtime_snapshot"))]
use once_cell::sync::Lazy;

#[cfg(not(feature = "dont_create_runtime_snapshot"))]
static COMPRESSED_RUNTIME_SNAPSHOT: &[u8] =
  include_bytes!(concat!(env!("OUT_DIR"), "/RUNTIME_SNAPSHOT.bin"));

#[cfg(not(feature = "dont_create_runtime_snapshot"))]
pub static RUNTIME_SNAPSHOT: Lazy<Box<[u8]>> = Lazy::new(
  #[allow(clippy::uninit_vec)]
  #[cold]
  #[inline(never)]
  || {
    let size =
      u32::from_le_bytes(COMPRESSED_RUNTIME_SNAPSHOT[0..4].try_into().unwrap())
        as usize;
    let mut vec = Vec::with_capacity(size);

    // SAFETY: vec is allocated with exact snapshot size (+ alignment)
    // SAFETY: non zeroed bytes are overwritten with decompressed snapshot
    unsafe {
      vec.set_len(size);
    }

    lzzzz::lz4::decompress(&COMPRESSED_RUNTIME_SNAPSHOT[4..], &mut vec)
      .unwrap();

    vec.into_boxed_slice()
  },
);

#[cfg(not(feature = "dont_create_runtime_snapshot"))]
pub fn deno_isolate_init() -> Snapshot {
  debug!("Deno isolate init with snapshots.");
  Snapshot::Static(&RUNTIME_SNAPSHOT)
}

#[cfg(not(feature = "include_js_files_for_snapshotting"))]
pub static SOURCE_CODE_FOR_99_MAIN_JS: &str = include_str!("js/99_main.js");

#[cfg(feature = "include_js_files_for_snapshotting")]
pub static PATH_FOR_99_MAIN_JS: &str =
  concat!(env!("CARGO_MANIFEST_DIR"), "/js/99_main.js");
