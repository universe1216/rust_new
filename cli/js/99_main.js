// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
import "ext:cli/runtime/js/99_main.js";
