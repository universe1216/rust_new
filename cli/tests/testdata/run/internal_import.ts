import "internal:runtime/01_errors.js";
