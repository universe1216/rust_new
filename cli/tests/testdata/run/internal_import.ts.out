error: Unsupported scheme "internal" for module "internal:runtime/01_build.js". Supported schemes: [
    "data",
    "blob",
    "file",
    "http",
    "https",
]
    at [WILDCARD]
